File: app/src/main/java/com/tcc/sicv/ui/ProfileActivity.java
package com.tcc.sicv.ui;

import android.arch.lifecycle.Observer;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import com.tcc.sicv.R;
import com.tcc.sicv.base.BaseActivity;
import com.tcc.sicv.data.model.FlowState;
import com.tcc.sicv.data.model.User;
import com.tcc.sicv.data.preferences.PreferencesHelper;
import com.tcc.sicv.presentation.ProfileViewModel;
import com.tcc.sicv.utils.OnAlertButtonClick;

public class ProfileActivity extends BaseActivity {
    private ProfileViewModel mViewModel;
    private TextView nameTv;
    private TextView cpfTv;
    private TextView emailTv;
    private TextView dateTv;
    private TextView telTv;
    private Button logoutBt;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_profile);
        setupToolbar(R.id.main_toolbar, R.string.user_profile, true);
        mViewModel = new ProfileViewModel(new PreferencesHelper(getApplication()));
        setupViews();
        creatingObservers();
    }

    private void creatingObservers() {
        mViewModel.getFlowState().observe(this, new Observer<FlowState<User>>() {
            @Override
            public void onChanged(@Nullable FlowState<User> flowState) {
                if (flowState != null) {
                    handleWithMainFlow(flowState);
                }
            }
        });
        mViewModel.getLogoutState().observe(this, new Observer<FlowState<Void>>() {
            @Override
            public void onChanged(@Nullable FlowState<Void> logoutState) {
                handleWithLogoutState(logoutState);
            }
        });
        logoutBt.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                createConfirmLogoutDialog(
                        getString(R.string.logout_message_confirm),
                        positiveListener,
                        negativeListener);
            }
        });
    }

    private void handleWithLogoutState(@Nullable FlowState<Void> logoutState) {
        if (logoutState != null) {
            switch (logoutState.getStatus()) {
                case LOADING:
                    showLoadingDialog();
                    break;
                case SUCCESS:
                    hideLoadingDialog();
                    Intent intent = new Intent(ProfileActivity.this, LoginActivity.class);
                    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
                    startActivity(intent);
                    break;
                case ERROR:
                    hideLoadingDialog();
                    handleErrors(logoutState.getThrowable());
                    break;
            }
        }
    }

    private void handleWithMainFlow(FlowState<User> flowState) {
        switch (flowState.getStatus()) {
            case LOADING:
                showLoadingDialog();
                logoutBt.setEnabled(false);
                mViewModel.getUserProfile();
                break;
            case SUCCESS:
                hideLoadingDialog();
                User userProfile = flowState.getData();
                if (userProfile == null) return;
                logoutBt.setEnabled(true);
                nameTv.setText(userProfile.getName());
                cpfTv.setText(userProfile.getCpf());
                emailTv.setText(userProfile.getEmail());
                dateTv.setText(userProfile.getDate());
                telTv.setText(userProfile.getTel());
                break;
            case ERROR:
                hideLoadingDialog();
                handleErrors(flowState.getThrowable());
                break;
        }
    }

    private void setupViews() {
        logoutBt = findViewById(R.id.logoutButton);
        nameTv = findViewById(R.id.nameTv);
        cpfTv = findViewById(R.id.cpfTv);
        emailTv = findViewById(R.id.emailTv);
        dateTv = findViewById(R.id.dateTv);
        telTv = findViewById(R.id.telephoneTv);
    }

    private OnAlertButtonClick positiveListener = new OnAlertButtonClick() {
        @Override
        public void onClickButton(DialogInterface dialog) { mViewModel.logout(); }
        @Override
        public String getText() { return getString(R.string.positive_button_text); }
    };

    private OnAlertButtonClick negativeListener = new OnAlertButtonClick() {
        public void onClickButton(DialogInterface dialog) { }
        @Override
        public String getText() { return getString(R.string.negative_button_text); }
    };
}

File: app/src/main/java/com/tcc/sicv/ui/ProfileActivity.kt
package com.tcc.sicv.ui

import android.arch.lifecycle.Observer
import android.content.Intent.FLAG_ACTIVITY_CLEAR_TASK
import android.content.Intent.FLAG_ACTIVITY_NEW_TASK
import android.os.Bundle
import com.tcc.sicv.R
import com.tcc.sicv.base.BaseActivity
import com.tcc.sicv.data.model.FlowState
import com.tcc.sicv.data.model.Status
import com.tcc.sicv.data.model.User
import com.tcc.sicv.data.preferences.PreferencesHelper
import com.tcc.sicv.presentation.ProfileViewModel
import com.tcc.sicv.utils.OnAlertButtonClick
import com.tcc.sicv.utils.startActivity
import kotlinx.android.synthetic.main.activity_profile.*

class ProfileActivity : BaseActivity() {
    private lateinit var mViewModel: ProfileViewModel

    private val positiveListener = object : OnAlertButtonClick {
        override val text: String
            get() = getString(R.string.positive_button_text)

        override fun onClickButton() {
            mViewModel.logout()
        }
    }

    private val negativeListener = object : OnAlertButtonClick {
        override val text: String
            get() = getString(R.string.negative_button_text)

        override fun onClickButton() {}
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_profile)
        setupToolbar(R.id.main_toolbar, R.string.user_profile, true)
        mViewModel = ProfileViewModel(PreferencesHelper(application))
        creatingObservers()
    }

    private fun creatingObservers() {
        mViewModel.getFlowState().observe(this, Observer { handleWithMainFlow(it) })
        mViewModel.getLogoutState().observe(this, Observer { handleWithLogoutState(it) })
        logoutButton.setOnClickListener {
            createDialog(
                    getString(R.string.logout_message_confirm),
                    positiveListener,
                    negativeListener)
        }

        refreshProfileLayout.setOnRefreshListener { mViewModel.getUserProfile() }
    }

    private fun handleWithLogoutState(logoutState: FlowState<Unit>?) {
        if (logoutState != null) {
            when (logoutState.status) {
                Status.LOADING -> refreshProfileLayout.isRefreshing = true
                Status.SUCCESS -> {
                    refreshProfileLayout.isRefreshing = false
                    startActivity<LoginActivity>(
                            flags = FLAG_ACTIVITY_NEW_TASK or FLAG_ACTIVITY_CLEAR_TASK
                    )
                }
                Status.ERROR -> {
                    refreshProfileLayout.isRefreshing = false
                    handleErrors(logoutState.throwable)
                }
                else -> {}
            }
        }
    }

    private fun handleWithMainFlow(flowState: FlowState<User>?) {
        when (flowState?.status) {
            Status.LOADING -> {
                refreshProfileLayout.isRefreshing = true
                logoutButton.isEnabled = false
            }
            Status.SUCCESS -> {
                refreshProfileLayout.isRefreshing = false
                handleWithSuccessMainFlow(flowState)
            }
            Status.ERROR -> {
                refreshProfileLayout.isRefreshing = false
                handleErrors(flowState.throwable)
            }
            else -> {}
        }
    }

    private fun handleWithSuccessMainFlow(flowState: FlowState<User>) {
        val userProfile = flowState.data ?: return
        logoutButton.isEnabled = true
        nameTv.text = userProfile.name
        cpfTv.text = userProfile.cpf
        emailTv.text = userProfile.email
        dateTv.text = userProfile.date
        telephoneTv.text = userProfile.tel
    }
}
